pkg/policy/policy_peer.go

```diff
@@ -91,3 +93,65 @@ func (rc L4RuleContext) PortProto() string {
 	port := strconv.Itoa(int(byteorder.NetworkToHost(uint16(rc.Port)).(uint16)))
 	return port + "/" + proto
 }
+
+// EntityContexts maps am entity to a L4RuleContexts
+type EntityContexts map[api.Entity]L4RuleContexts
+
+// DeepCopy returns a deep copy of EntityContexts
+func (ec EntityContexts) DeepCopy() EntityContexts {
+	cpy := make(EntityContexts)
+	for k, v := range ec {
+		cpy[k] = v.DeepCopy()
+	}
+	return cpy
+}
+
+// NewEntityContexts returns a new L4RuleContexts created.
+func NewEntityContexts() EntityContexts {
+	return EntityContexts(make(map[api.Entity]L4RuleContexts))
+}
+
+// CIDRContexts maps a CIDR to a L4RuleContexts
+type CIDRContexts map[api.CIDR]L4RuleContexts
+
+// DeepCopy returns a deep copy of EntityContexts
+func (cc CIDRContexts) DeepCopy() CIDRContexts {
+	cpy := make(CIDRContexts)
+	for k, v := range cc {
+		cpy[k] = v.DeepCopy()
+	}
+	return cpy
+}
+
+// NewCIDRContexts returns a new L4RuleContexts created.
+func NewCIDRContexts() CIDRContexts {
+	return CIDRContexts(make(map[api.CIDR]L4RuleContexts))
+}
+
+// Policy contains all policies combined by each type of policy peer.
```

Comment:
Can you describe the relationship between `Policy` and `Repository` here?

What I have in mind is that the `Repository` is the representation of user policy created from the API. It should always match exactly what was configured. It is desired state. When endpoint regeneration occurs, a goroutine iterates through the `Repository` and generates this `Policy`, which should be an immutable translation of the `Repository` which organises the policy in a more datapath-friendly manner. Later on, we can add a `pkg/datapath` which will accept this `Policy` and creates actual maps and BPF progs which implement this policy, then as a final step the regeneration logic can trigger the atomic swap.

Reply:
That is spot on!